iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp

```diff
@@ -28,30 +29,84 @@ namespace iox
+                                          cxx::newtype::MoveConstructable>
+{
+    using ThisType::ThisType;
+};
+
+using segment_id_underlying_t = typename segment_id_t::value_type;
+
+/// @brief pointer class to use when pointer and pointee are located in different shared memory segments
+/// We can have the following scenario:
+/// Pointer p is stored in segment S1 and points to object X of type T in segment S2.
+///
+/// Shared Memory   S1:  p              S2:  X
+///                      |___________________^
+/// App1            a1   b1             c1   d1
+/// App2            a2   b2             c2   d2
+///
+/// Now it is no longer true in general that both segments will be offset by the same difference in App2 and therefore
+/// relocatable pointers are no longer sufficient.
+/// Relative pointers solve this problem by incorporating the information from where they need to measure differences
+/// (i.e. relative to the given address). This requires an additional registration mechanism to be used by all
+/// applications where the start addresses and the size of all segments to be used are registered. Since these start
+/// address may differ between applications, each segment is identified by a unique id, which can be provided upon
+/// registration by the first application. In the figure, this means that the starting addresses of both segments(a1, a2
+/// and c1, c2) would have to be registered in both applications.
+/// Once this registration is done, relative pointers can be constructed from raw pointers similar to relocatable
+/// pointers.
+/// @note It should be noted that relocating a memory segment will invalidate relative pointers, i.e. relative pointers
+/// are NOT relocatable. This is because the registration mechanism cannot be automatically informed about the copy of a
+/// whole segment, such a segment would have to be registered on its own (and the original segment deregistered).
 template <typename T>
-class RelativePointer : public BaseRelativePointer
+class RelativePointer
 {
   public:
     using const_ptr_t = const T*;
     using ptr_t = T*;
+    using offset_t = std::uintptr_t;
 
     /// @brief default constructs a RelativePointer as a logical nullptr
     RelativePointer() noexcept = default;
 
+    ~RelativePointer() noexcept = default;
+
     /// @brief constructs a RelativePointer pointing to the same pointee as ptr in a segment identified by id
     /// @param[in] ptr the pointer whose pointee shall be the same for this
     /// @param[in] id is the unique id of the segment
-    RelativePointer(ptr_t ptr, id_t id) noexcept;
+    RelativePointer(const ptr_t ptr, const segment_id_t id) noexcept;
 
     /// @brief constructs a RelativePointer from a given offset and segment id
     /// @param[in] offset is the offset
     /// @param[in] id is the unique id of the segment
-    RelativePointer(offset_t offset, id_t id) noexcept;
+    RelativePointer(const offset_t offset, const segment_id_t id) noexcept;
```

Comment:
@MatthiasKillat I see your point. This was discussed before and concluded that `NewType`, should be consistently used everywhere where you can easily swap the parameters by mistake. Not something vor `v3.0.` but `v4.0`.

I can't be private because of `ShmSafeUnmanagedChunk`. I fear we will not further refactor this for `v3.0`. API-wise it's ok from my side.

Reply:
This ctor is only used internally for now anyway (but of course everyone can use it). I would like to get it out of the public API and return to primitive types...